session/session.go

```diff
@@ -2946,7 +2946,13 @@ func (s *session) GetInfoSchema() sessionctx.InfoschemaMetaVersion {
 			return is
 		}
 	}
-	return domain.GetDomain(s).InfoSchema()
+
+	is := domain.GetDomain(s).InfoSchema()
```

Comment:
If user `set autocommit=0` or execute sql `begin` to start a txn, temporary table seems to be invisible. May be we should wrapper all returns when local temporary table exists event tidb_snapshot is set

Reply:
> If user set autocommit=0 or execute sql begin to start a txn, temporary table seems to be invisible.

In mysql 8.0:

```
mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 8.0.25-0ubuntu0.20.04.1 |
+-------------------------+
1 row in set (0.00 sec)

mysql> create temporary table tmp (id int);
Query OK, 0 rows affected (0.00 sec)

mysql> select * from tmp;
Empty set (0.00 sec)

mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from tmp;
Empty set (0.00 sec)

mysql> set autocommit = 0;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from tmp;
Empty set (0.00 sec)
```

Reply:
Is it the same with MySQL in your PR? It seems when it's in a transaction, the returned infoschema is `InfoSchema` instead of `TemporaryTableAttachedInfoSchema`.

Reply:
Yes, the behaviour is the same here.

The first time we start a transaction, the  snapshot of `InfoSchema` is taken.
During the transaction we can not use a different `InfoSchema`.

Reply:
Temporary table should be visible in transaction and that is a correct behavior. I ignored your modify in PrepareTxnCtx, so temp tables are visible in txn in most cases. But some cases will lost temporary table for this pr:

for example in stale read txn
```
mysql> create temporary table db1.tmp1 (id int primary key);
Query OK, 0 rows affected (0.05 sec)

mysql> START TRANSACTION READ ONLY AS OF TIMESTAMP NOW(3);
Query OK, 0 rows affected (0.01 sec)

mysql> select * from db1.tmp1;
ERROR 1146 (42S02): Table 'db1.tmp1' doesn't exist
```

Reply:
I think the temporary table behaviour related to the snapshot is undefined.
Because the temporary table can not support MVCC, and there is no historial version data.

```
select current_ts(); --- t1
create temporary table t (id int);

set @@snapshot = t1;
select * from t    -- t exist or not?

set @@snapshot = now()
select * from t    -- t exist or not?
```

Reply:
Done